File: app/model/Enums.scala
package model

import enumeratum.{EnumEntry, PlayEnum}
import sangria.macros.derive.deriveEnumType

sealed trait Importance extends EnumEntry

object Importance extends PlayEnum[Importance] {

  case object Less   extends Importance
  case object Medium extends Importance
  case object More   extends Importance

  override def values = findValues
  val graphQLType     = deriveEnumType[Importance]()
}

sealed trait AnnotationType extends EnumEntry

object AnnotationType extends PlayEnum[AnnotationType] {

  case object Manual            extends AnnotationType
  case object Automatic         extends AnnotationType
  case object RejectedAutomatic extends AnnotationType

  val values: IndexedSeq[AnnotationType] = findValues
  val graphQLType                        = deriveEnumType[AnnotationType]()
}

sealed trait Applicability extends EnumEntry

object Applicability extends PlayEnum[Applicability] {

  case object NotSpecified  extends Applicability
  case object NotApplicable extends Applicability
  case object Applicable    extends Applicability

  override def values = findValues
  val graphQLType     = deriveEnumType[Applicability]()
}

// Correctness

sealed trait Correctness extends EnumEntry

object Correctness extends PlayEnum[Correctness] {
  case object Correct     extends Correctness
  case object Partially   extends Correctness
  case object Wrong       extends Correctness
  case object Unspecified extends Correctness

  def values      = findValues
  val graphQLType = deriveEnumType[Correctness]()
}

// ErrorType

sealed trait ErrorType extends EnumEntry

case object ErrorType extends PlayEnum[ErrorType] {

  case object Neutral extends ErrorType
  case object Missing extends ErrorType
  case object Wrong   extends ErrorType

  override def values = findValues
  val graphQLType     = deriveEnumType[ErrorType]()
}

sealed trait MatchStatus extends EnumEntry

object MatchStatus extends PlayEnum[MatchStatus] {

  case object Automatic extends MatchStatus
  case object Manual    extends MatchStatus
  case object Deleted   extends MatchStatus

  override def values = findValues
  val graphQLType     = deriveEnumType[MatchStatus]()
}

sealed trait Rights extends EnumEntry

object Rights extends PlayEnum[Rights] {

  case object Student   extends Rights
  case object Corrector extends Rights
  case object Admin     extends Rights

  override def values = findValues
  val graphQLType     = deriveEnumType[Rights]()
}

File: vite_ui/src/model/enums.ts
import { Applicability, ErrorType, Importance } from '../graphql';

export function stringifyApplicability(a: Applicability): string {
  return {
    [Applicability.NotSpecified]: '',
    [Applicability.NotApplicable]: '(-)',
    [Applicability.Applicable]: '(+)'
  }[a];
}

export const errorTypes: ErrorType[] = [ErrorType.Neutral, ErrorType.Missing, ErrorType.Wrong];

export function borderColorForErrorType(errorType: ErrorType): string {
  return {
    [ErrorType.Neutral]: 'border-slate-500',
    [ErrorType.Wrong]: 'border-red-500',
    [ErrorType.Missing]: 'border-amber-500',
  }[errorType];
}

export const importanceTypes: Importance[] = [Importance.Less, Importance.Medium, Importance.More];

export function importanceSymbol(importance: Importance): string {
  return {
    [Importance.Less]: '  *',
    [Importance.Medium]: ' **',
    [Importance.More]: '***'
  }[importance];
}

export function importanceFontStyle(importance: Importance): string {
  return {
    [Importance.Less]: 'italic',
    [Importance.Medium]: '',
    [Importance.More]: 'font-bold'
  }[importance];
}
